File: C#/BluffinPokerServer/ServerClientLobby.cs
using System;
using System.Collections.Generic;
using System.Text;
using EricUtility.Networking.Commands;
using Com.Ericmas001.Game.Poker.Protocol.Server;
using System.Net.Sockets;
using Com.Ericmas001.Game.Poker.Protocol.Commands.Lobby;
using System.IO;
using PokerWorld.Game;
using Com.Ericmas001.Game.Poker.Protocol.Commands.Lobby.Training;
using Com.Ericmas001.Game.Poker.Persistance;
using Com.Ericmas001.Game.Poker.Protocol.Commands.Lobby.Career;
using EricUtility;
using System.Web;
using Com.Ericmas001.Game.Poker.Protocol.Commands;
using Com.Ericmas001.Game.Poker.DataTypes.Enums;
using PokerWorld.Game.Rules;


namespace BluffinPokerServer
{
    public class ServerClientLobby : CommandTCPCommunicator<LobbyServerCommandObserver>
    {
        private string m_PlayerName = "?";
        private readonly ServerLobby m_Lobby;
        Dictionary<int, GameServer> m_Tables = new Dictionary<int, GameServer>();

        public ServerClientLobby(TcpClient client, ServerLobby lobby)
            : base(client)
        {
            m_Lobby = lobby;
        }
        protected override void InitializeCommandObserver()
        {
            m_CommandObserver.CommandReceived += m_CommandObserver_CommandReceived;
            m_CommandObserver.DisconnectCommandReceived += m_CommandObserver_DisconnectCommandReceived;
            m_CommandObserver.ListTableCommandReceived += m_CommandObserver_ListTableCommandReceived;
            m_CommandObserver.JoinTableCommandReceived += m_CommandObserver_JoinTableCommandReceived;
            m_CommandObserver.GameCommandReceived += m_CommandObserver_GameCommandReceived;
            m_CommandObserver.SupportedRulesCommandReceived += m_CommandObserver_SupportedRulesCommandReceived;
            m_CommandObserver.CreateTableCommandReceived += m_CommandObserver_CreateTableCommandReceived;

            //Training
            m_CommandObserver.IdentifyCommandReceived += m_CommandObserver_IdentifyCommandReceived;
            
            //Career
            m_CommandObserver.CheckDisplayExistCommandReceived += m_CommandObserver_CheckDisplayExistCommandReceived;
            m_CommandObserver.CheckUserExistCommandReceived += m_CommandObserver_CheckUserExistCommandReceived;
            m_CommandObserver.CreateUserCommandReceived += m_CommandObserver_CreateUserCommandReceived;
            m_CommandObserver.AuthenticateUserCommandReceived += m_CommandObserver_AuthenticateUserCommandReceived;
            m_CommandObserver.GetUserCommandReceived += m_CommandObserver_GetUserCommandReceived;
        }

        void m_CommandObserver_CreateTableCommandReceived(object sender, CommandEventArgs<CreateTableCommand> e)
        {
            Send(e.Command.EncodeResponse(CreateTable(e.Command)));
        }

        void m_CommandObserver_SupportedRulesCommandReceived(object sender, CommandEventArgs<SupportedRulesCommand> e)
        {
            Send(e.Command.EncodeResponse());
        }

        private int CreateTable(CreateTableCommand c)
        {
            int res = m_Lobby.CreateTable(c);
            LogManager.Log(LogLevel.Message, "ServerClientLobby.m_CommandObserver_{3}Received", "> Client '{0}' {3}: {2}:{1}", m_PlayerName, c.GameRules.TableName, res, c.GameRules.CurrentLobby.LobbyType);
            return res;
        }

        void m_CommandObserver_GetUserCommandReceived(object sender, CommandEventArgs<GetUserCommand> e)
        {
            UserInfo u = DataManager.Persistance.Get(e.Command.Username);
            Send(e.Command.EncodeResponse(u.Email,u.DisplayName,u.TotalMoney));
        }

        void m_CommandObserver_CheckDisplayExistCommandReceived(object sender, CommandEventArgs<CheckDisplayExistCommand> e)
        {
            Send(e.Command.EncodeResponse(m_Lobby.NameUsed(e.Command.DisplayName) || DataManager.Persistance.IsDisplayNameExist(e.Command.DisplayName)));
        }

        void m_CommandObserver_AuthenticateUserCommandReceived(object sender, CommandEventArgs<Com.Ericmas001.Game.Poker.Protocol.Commands.Lobby.Career.AuthenticateUserCommand> e)
        {
            UserInfo u = DataManager.Persistance.Authenticate(e.Command.Username, e.Command.Password);

            if (u != null)
                m_PlayerName = u.DisplayName;

            bool ok = (u != null && !m_Lobby.NameUsed(m_PlayerName));
            if( ok )
                m_Lobby.AddName(m_PlayerName);

            LogManager.Log(LogLevel.Message, "ServerClientLobby.m_CommandObserver_AuthenticateUserCommandReceived", "> Client authenticate to Career Server as : {0}. Success={1}", m_PlayerName, ok);
            Send(e.Command.EncodeResponse(ok));
        }

        void m_CommandObserver_CreateUserCommandReceived(object sender, CommandEventArgs<Com.Ericmas001.Game.Poker.Protocol.Commands.Lobby.Career.CreateUserCommand> e)
        {
            CreateUserCommand c = e.Command;
            bool ok = !DataManager.Persistance.IsUsernameExist(c.Username) && !DataManager.Persistance.IsDisplayNameExist(e.Command.DisplayName);

            if( ok)
                DataManager.Persistance.Register(new UserInfo(c.Username,c.Password,c.Email,c.DisplayName,7500));

            LogManager.Log(LogLevel.Message, "ServerClientLobby.m_CommandObserver_CreateUserCommandReceived", "> Client register to Career Server as : {0}. Success={1}", c.Username, ok);
            Send(e.Command.EncodeResponse(ok));
        }

        void m_CommandObserver_CheckUserExistCommandReceived(object sender, CommandEventArgs<Com.Ericmas001.Game.Poker.Protocol.Commands.Lobby.Career.CheckUserExistCommand> e)
        {
            Send(e.Command.EncodeResponse(DataManager.Persistance.IsUsernameExist(e.Command.Username)));
        }

        void m_CommandObserver_GameCommandReceived(object sender, CommandEventArgs<GameCommand> e)
        {
            m_Tables[e.Command.TableID].Incoming(e.Command.DecodedCommand);
        }

        void m_CommandObserver_JoinTableCommandReceived(object sender, CommandEventArgs<JoinTableCommand> e)
        {
            GameServer client = null;
            PokerGame game = m_Lobby.GetGame(e.Command.TableID);
            PokerTable table = game.Table;

            if (game.Rules.CurrentLobby.LobbyType == LobbyTypeEnum.Training)
                client = new GameServer(e.Command.TableID, game, m_PlayerName, ((LobbyOptionsTraining)game.Rules.CurrentLobby).StartingAmount);
            else
                client = new GameServer(e.Command.TableID, game, DataManager.Persistance.Get(e.Command.PlayerName));

            client.LeftTable += new EventHandler<EricUtility.KeyEventArgs<int>>(client_LeftTable);
            client.SendedSomething += new EventHandler<EricUtility.KeyEventArgs<string>>(client_SendedSomething);

            if (!game.IsRunning)
            {
                Send(e.Command.EncodeErrorResponse());
                return;
            }

            // Verify the player does not already playing on that table.
            if (!table.ContainsPlayer(e.Command.PlayerName))
            {
                bool ok = client.JoinGame();
                if (!ok)
                    Send(e.Command.EncodeErrorResponse());
                else
                {
                    m_Tables.Add(e.Command.TableID, client);
                    client.Start();

                    LogManager.Log(LogLevel.Message, "ServerClientLobby.m_CommandObserver_JoinTableCommandReceived", "> Client '{0}' seated ({3}) at table: {2}:{1}", m_PlayerName, table.Rules.TableName, e.Command.TableID, client.Player.NoSeat);
                    Send(e.Command.EncodeResponse(client.Player.NoSeat));

                    client.SitIn();
                }
            }
            else
            {
                Send(e.Command.EncodeErrorResponse());
            }
        }

        void client_LeftTable(object sender, EricUtility.KeyEventArgs<int> e)
        {
            m_Tables.Remove(e.Key);
        }

        void client_SendedSomething(object sender, EricUtility.KeyEventArgs<string> e)
        {
            GameServer client = (GameServer)sender;
            Send(new GameCommand() { TableID = client.ID, EncodedCommand = e.Key });
        }
        public override void OnReceiveCrashed(Exception e)
        {
            if (e is IOException)
            {
                LogManager.Log(LogLevel.Error, "ServerClientLobby.OnReceiveCrashed", "Server lost connection with {0}", m_PlayerName);
                m_Lobby.RemoveName(m_PlayerName);
                Close();
            }
            else
                base.OnReceiveCrashed(e);
        }

        protected override void Send(string line)
        {
            LogManager.Log(LogLevel.MessageVeryLow, "ServerClientLobby.Send", "Server SEND to {0} [{1}]", m_PlayerName, line);
            base.Send(line);
        }

        void m_CommandObserver_CommandReceived(object sender, StringEventArgs e)
        {
            LogManager.Log(LogLevel.MessageVeryLow, "ServerClientLobby.m_CommandObserver_CommandReceived", "Server RECV from {0} [{1}]", m_PlayerName, e.Str);
        }

        void m_CommandObserver_IdentifyCommandReceived(object sender, CommandEventArgs<IdentifyCommand> e)
        {
            IdentifyCommand c = e.Command;
            m_PlayerName = c.Name;
            bool ok = !m_Lobby.NameUsed(m_PlayerName) && !DataManager.Persistance.IsDisplayNameExist(m_PlayerName);
            LogManager.Log(LogLevel.Message, "ServerClientLobby.m_CommandObserver_IdentifyCommandReceived", "> Client indentifying training server as : {0}. Success={1}", m_PlayerName, ok);
            Send(c.EncodeResponse(ok));
            if (ok)
                m_Lobby.AddName(m_PlayerName);
        }

        void m_CommandObserver_DisconnectCommandReceived(object sender, CommandEventArgs<DisconnectCommand> e)
        {
            LogManager.Log(LogLevel.Message, "ServerClientLobby.m_CommandObserver_DisconnectCommandReceived", "> Client disconnected: {0}", m_PlayerName);
            DisconnectCommand c = e.Command;
            m_Lobby.RemoveName(m_PlayerName);
            Close();
        }

        void m_CommandObserver_ListTableCommandReceived(object sender, CommandEventArgs<ListTableCommand> e)
        {
            ListTableCommand c = e.Command;
            Send(c.EncodeResponse(m_Lobby.ListTables(c.LobbyTypes)));
        }
    }
}

File: trunk/JAVA/BluffinServer/src/server/ServerClientLobby.java
package server;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.net.SocketException;
import java.util.Collections;
import java.util.Map;
import java.util.TreeMap;

import bluffinmuffin.poker.game.PokerGame;
import bluffinmuffin.poker.game.TableInfo;
import bluffinmuffin.protocol.GameTCPServer;
import bluffinmuffin.protocol.commands.DisconnectCommand;
import bluffinmuffin.protocol.commands.ICommand;
import bluffinmuffin.protocol.commands.lobby.CreateTableCommand;
import bluffinmuffin.protocol.commands.lobby.GameCommand;
import bluffinmuffin.protocol.commands.lobby.IdentifyCommand;
import bluffinmuffin.protocol.commands.lobby.JoinTableCommand;
import bluffinmuffin.protocol.commands.lobby.ListTableCommand;
import bluffinmuffin.protocol.observer.lobby.LobbyServerAdapter;
import bluffinmuffin.protocol.observer.lobby.LobbyServerObserver;


/**
 * This class represents a client for ServerLobby.
 */
public class ServerClientLobby extends Thread
{
    private String m_playerName = "?";
    private final ServerLobby m_lobby;
    private final LobbyServerObserver m_commandObserver = new LobbyServerObserver();
    Map<Integer, GameTCPServer> m_tables = Collections.synchronizedMap(new TreeMap<Integer, GameTCPServer>());
    
    // Communications with the client
    private Socket m_socket = null;
    private PrintWriter m_toClient;
    private BufferedReader m_fromClient;
    
    public ServerClientLobby(Socket socket, ServerLobby lobby)
    {
        m_socket = socket;
        m_lobby = lobby;
        try
        {
            m_toClient = new PrintWriter(socket.getOutputStream(), true /* autoFlush */);
            m_fromClient = new BufferedReader(new InputStreamReader(socket.getInputStream()));
        }
        catch (final IOException e)
        {
            e.printStackTrace();
        }
    }
    
    public boolean isConnected()
    {
        return m_socket.isConnected() && !m_socket.isClosed();
    }
    
    /**
     * Send message to client.
     * 
     * @param p_msg
     *            - Message to send.
     */
    public void sendMessage(String p_msg)
    {
        System.out.println("Server SEND to " + m_playerName + " [" + p_msg + "]");
        m_toClient.println(p_msg);
    }
    
    protected String receive() throws IOException
    {
        final String line = m_fromClient.readLine();
        m_commandObserver.receiveSomething(line);
        return line;
    }
    
    public void send(ICommand p_msg)
    {
        sendMessage(p_msg.encodeCommand());
    }
    
    @Override
    public void run()
    {
        initializeCommandObserver();
        while (isConnected())
        {
            try
            {
                receive();
            }
            catch (final SocketException e)
            {
                System.out.println("Connection lost with " + m_playerName);
                m_lobby.removeName(m_playerName);
                return;
            }
            catch (final Exception e)
            {
                e.printStackTrace();
            }
            
        }
    }
    
    private void initializeCommandObserver()
    {
        m_commandObserver.subscribe(new LobbyServerAdapter()
        {
            @Override
            public void commandReceived(String command)
            {
                System.out.println("Server RECV from " + m_playerName + " [" + command + "]");
            }
            
            @Override
            public void connectCommandReceived(IdentifyCommand command)
            {
                m_playerName = command.getPlayerName();
                final boolean ok = !m_lobby.isNameUsed(m_playerName);
                sendMessage(command.encodeResponse(ok));
                if (ok)
                {
                    m_lobby.addName(m_playerName);
                }
            }
            
            @Override
            public void disconnectCommandReceived(DisconnectCommand command)
            {
                m_lobby.removeName(m_playerName);
                try
                {
                    m_toClient.close();
                    m_fromClient.close();
                    m_socket.close();
                }
                catch (final IOException e)
                {
                    e.printStackTrace();
                }
            }
            
            @Override
            public void createTableCommandReceived(CreateTableCommand command)
            {
                sendMessage(command.encodeResponse(m_lobby.createTable(command)));
            }
            
            @Override
            public void listTableCommandReceived(ListTableCommand command)
            {
                sendMessage(command.encodeResponse(m_lobby.listTables()));
            }
            
            @Override
            public void joinTableCommandReceived(JoinTableCommand command)
            {
                final GameTCPServer client = new GameTCPServer(m_lobby.getGame(command.getTableID()), m_playerName, 1500);
                final PokerGame game = client.getGame();
                final TableInfo table = game.getTable();
                if (!game.isRunning())
                {
                    sendMessage(command.encodeErrorResponse());
                    return;
                }
                
                // Verify the player does not already playing on that table.
                if (!table.containsPlayer(command.getPlayerName()))
                {
                    final boolean ok = client.joinGame();
                    if (!ok)
                    {
                        sendMessage(command.encodeErrorResponse());
                    }
                    else
                    {
                        m_tables.put(command.getTableID(), client);
                        client.start();
                        sendMessage(command.encodeResponse(client.getPlayer().getNoSeat()));
                        client.sitIn();
                        new ServerClientSender(command.getTableID(), client, ServerClientLobby.this).start();
                    }
                }
                else
                {
                    sendMessage(command.encodeErrorResponse());
                }
            }
            
            @Override
            public void gameCommandReceived(GameCommand command)
            {
                try
                {
                    m_tables.get(command.getTableId()).incoming(command.getCommand());
                }
                catch (final InterruptedException e)
                {
                    e.printStackTrace();
                }
            }
        });
        
    }
}
